Meta.meta on synhetic module throws dataflow error

hubertp opened the pull request:
### Pull Request Description

It appears that under some undetermined scenarios synthetic modules didn't have an IR. Therefore calling any methods on them would lead to NPE.
Encountered the crash in the wild:
```
[org.enso.interpreter.instrument.job.ProgramExecutionSupport$] Execution of function main failed (Cannot invoke "org.enso.compiler.core.ir.Module.isPrivate()" because "this.ir" is null).
java.lang.NullPointerException: Cannot invoke "org.enso.compiler.core.ir.Module.isPrivate()" because "this.ir" is null
    at org.enso.runtime/org.enso.interpreter.runtime.Module.isPrivate(Module.java:248)
    at org.enso.runtime/org.enso.interpreter.node.expression.builtin.meta.MetaKindNode.execute(MetaKindNode.java:34)
    at org.enso.runtime/org.enso.interpreter.node.expression.builtin.meta.MetaKindMethodGen.handleExecute(MetaKindMethodGen.java:114)
    at 
```

This change enforces that synthetic modules are always private, irrespective of whether they have an IR or not.
This revealed some violations in our own stdlib which had to be workaround by creating stub modules.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

hubertp committed:
Meta.meta on synhetic module throws dataflow error

Synthetic modules don't have an IR. Therefore calling any methods on
them would lead to NPE.

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java:
```diff
@@ -245,7 +245,7 @@ public boolean isSynthetic() {
    */
   public boolean isPrivate() {
-    return ir.isPrivate();
+    return ir == null || ir.isPrivate();
```
Why should `isPrivate` be `true` for synthetic modules?

(Not saying it shouldn't but I would like understand the rationale)

hubertp replied:
Well.. they are either private or not. And I don't see any use-cases for making them public, especially given that there is no source code/documentation available for them for the users and we don't give any guarantees how those modules are being constructed.

I believe @Akirathan's had a similar intuition that they should be private.

Akirathan replied:
Rule of a thumb is: The more stuff we can make private, the better. Unless something breaks. And _synthetic_ modules certainly feel like they should be private.

JaroslavTulach reviewed (approved):
Thank you for the NPE fix and for the reasoning about privateness: 
- private is good defensive default
- more stuff private is good, if things work 
- making public something that cannot have documentation is wrong
    - btw. Java did this mistake
    - missing constructor implies public default one
    - and it cannot have documentation

hubertp committed:
Merge branch 'develop' into wip/hubert/guard-npe

hubertp committed:
Merge branch 'develop' into wip/hubert/guard-npe

hubertp commented:
There are issues with NI in tests - somehow in NI the IR is being set for synthetic modules. Remains to be resolved...

hubertp committed:
Make synethic modules always private

Previously, module was private if it had no IR or IR was private only.

hubertp committed:
Workaround restriction on private synthetic module

Previously, we allowed, unintentionally, synthetic modules to be public.
As this PR ensures that synthetic modules are private it revealed two
cases where we violated our own restriction. By adding stub modules that
will have implicitly added synthetic imports/exports, we workaround the
restriction.

hubertp committed:
Fix restriction in tests

hubertp committed:
update API

hubertp committed:
Merge branch 'develop' into wip/hubert/guard-npe

hubertp committed:
Adapt test
